pysindy/utils/axes.py

```diff
@@ -0,0 +1,187 @@
+
+    @property
+    def n_sample(self):
+        return self.shape[self.ax_sample] if self.ax_sample is not None else 1
+
+    @property
+    def n_coord(self):
+        return self.shape[self.ax_coord] if self.ax_coord is not None else 1
+
+    def __array_ufunc__(
+        self, ufunc, method, *inputs, out=None, **kwargs
+    ):  # this method is called whenever you use a ufunc
+        args = []
+        for input_ in inputs:
+            if isinstance(input_, AxesArray):
+                args.append(input_.view(np.ndarray))
+            else:
+                args.append(input_)
+
+        outputs = out
+        if outputs:
+            out_args = []
+            for output in outputs:
+                if isinstance(output, AxesArray):
+                    out_args.append(output.view(np.ndarray))
+                else:
+                    out_args.append(output)
+            kwargs["out"] = tuple(out_args)
+        else:
+            outputs = (None,) * ufunc.nout
+        results = super().__array_ufunc__(ufunc, method, *args, **kwargs)
+        if results is NotImplemented:
+            return NotImplemented
+        if method == "at":
+            return
+        if ufunc.nout == 1:
+            results = (results,)
+        results = tuple(
+            (AxesArray(np.asarray(result), self.__dict__) if output is None else output)
+            for result, output in zip(results, outputs)
+        )
+        return results[0] if len(results) == 1 else results
+
+    def __array_function__(self, func, types, args, kwargs):
+        if func not in HANDLED_FUNCTIONS:
+            arr = super(AxesArray, self).__array_function__(func, types, args, kwargs)
+            if isinstance(arr, np.ndarray):
+                return AxesArray(arr, axes=self.__dict__)
+            elif arr is not None:
+                return arr
+            return
+        if not all(issubclass(t, AxesArray) for t in types):
+            return NotImplemented
+        return HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+
+def implements(numpy_function):
+    """Register an __array_function__ implementation for MyArray objects."""
+
+    def decorator(func):
+        HANDLED_FUNCTIONS[numpy_function] = func
+        return func
+
+    return decorator
+
+
+@implements(np.concatenate)
+def concatenate(arrays, axis=0):
+    parents = [np.asarray(obj) for obj in arrays]
+    ax_list = [obj.__dict__ for obj in arrays if isinstance(obj, AxesArray)]
+    for ax1, ax2 in zip(ax_list[:-1], ax_list[1:]):
+        if ax1 != ax2:
+            raise TypeError("Concatenating >1 AxesArray with incompatible axes")
+    return AxesArray(np.concatenate(parents, axis), axes=ax_list[0])
+
+
+def comprehend_axes(x):
+    axes = {}
+    axes["ax_coord"] = len(x.shape) - 1
+    axes["ax_time"] = len(x.shape) - 2
+    axes["ax_spatial"] = list(range(len(x.shape) - 2))
+    return axes
+
+
+def ax_time_to_ax_sample(x: AxesArray) -> AxesArray:
```

Comment:
Is this function actually used for anything? It is applied in some but not all feature library transforms and in the SINDy score/predict. But the SampleConcatter implements a more general version of converting axes other axes to samples (which is why it doesn't matter that some feature libraries don't call ax_time_to_samples). I'd probably try to remove this, let feature libraries keep the axes unchanged, and use the SampleConcatter everywhere instead. But this is not pressing, so could be deferred.

Reply:
I'm happy to remove it - the function was created at a time when `drop_nan_samples` and `drop_random_samples` were called before the feature libraries `transform()`, and therefore before samples were concatted.  Now it probably can be removed.

Reply:
Removed in new commit